src/io/flutter/inspector/InspectorGroupManagerService.java

```diff
@@ -0,0 +1,319 @@
+     */
+    default void requestRepaint(boolean force) {
+    }
+
+    /**
+     * Called whenever Flutter renders a frame.
+     */
+    default void onFlutterFrame() {
+    }
+
+    default void notifyAppReloaded() {
+    }
+
+    default void notifyAppRestarted() {
+    }
+
+    /**
+     * Called when the inspector selection has changed.
+     */
+    default void onSelectionChanged(DiagnosticsNode selection) {
+    }
+  }
+
+  private final Set<Listener> listeners = new HashSet<>();
+  boolean started = false;
+  private CompletableFuture<InspectorService> inspectorServiceFuture;
+  private FlutterApp app;
+  private FlutterApp.FlutterAppListener appListener;
+  private DiagnosticsNode selection;
+  private InspectorService inspectorService;
+  private InspectorObjectGroupManager selectionGroups;
+
+  public InspectorGroupManagerService(Project project) {
+    FlutterAppManager.getInstance(project).getActiveAppAsStream().listen(
+      this::updateActiveApp, true);
+    Disposer.register(project, this);
+  }
+
+  @NotNull
+  public static InspectorGroupManagerService getInstance(@NotNull final Project project) {
+    return ServiceManager.getService(project, InspectorGroupManagerService.class);
+  }
+
+  public InspectorService getInspectorService() {
+    return inspectorService;
+  }
+
+  public void addListener(@NotNull Listener listener, Disposable disposable) {
+    synchronized (listeners) {
+      listeners.add(listener);
+    }
+    // Update the listener with the current active state if any.
+    if (inspectorService != null) {
+      listener.onInspectorAvailable(inspectorService);
+    }
+    if (selection != null) {
+      listener.onSelectionChanged(selection);
+    }
+    Disposer.register(disposable, () -> removeListener(listener));
+  }
+
+  private void removeListener(@NotNull Listener listener) {
+    synchronized (listeners) {
+      listeners.remove(listener);
+    }
+  }
+
+  private void loadSelection() {
+    selectionGroups.cancelNext();
+
+    final CompletableFuture<DiagnosticsNode> pendingSelectionFuture =
+      selectionGroups.getNext().getSelection(null, InspectorService.FlutterTreeType.widget, true);
+    selectionGroups.getNext().safeWhenComplete(pendingSelectionFuture, (selection, error) -> {
+      if (error != null) {
+        //                  FlutterUtils.warn(LOG, error);
+        selectionGroups.cancelNext();
+        return;
+      }
+
+      selectionGroups.promoteNext();
+      invokeOnAllListeners((listener) -> {
+        listener.onSelectionChanged(selection);
+      });
+    });
+  }
+
+  private void requestRepaint(boolean force) {
+    invokeOnAllListeners((listener) -> {
+      listener.requestRepaint(force);
+    });
+  }
+
+  private void updateActiveApp(FlutterApp app) {
+    if (app == this.app) {
+      return;
+    }
+    selection = null;
+
+    if (this.app != null && appListener != null) {
+      this.app.removeStateListener(appListener);
+      this.app = null;
```

Comment:
this line can be removed since `this.app` is set to `app` directly below this if block

Reply:
done